src/Assistant.ts

```diff
@@ -0,0 +1,387 @@
+      threadContextStore = new DefaultThreadContextStore(),
+      threadStarted,
+      // When `threadContextChanged` method is not provided, fallback to
+      // AssistantContextStore's save method. If a custom store has also not
+      // been provided, the default save context-via-metadata approach is used.
+      // See DefaultThreadContextStore for details of this implementation.
+      threadContextChanged = (args) => threadContextStore.save(args),
+      userMessage,
+    } = config;
+
+    this.threadContextStore = threadContextStore;
+    this.threadStarted = Array.isArray(threadStarted) ? threadStarted : [threadStarted];
+    this.threadContextChanged = Array.isArray(threadContextChanged) ? threadContextChanged : [threadContextChanged];
+    this.userMessage = Array.isArray(userMessage) ? userMessage : [userMessage];
+  }
+
+  public getMiddleware(): Middleware<AnyMiddlewareArgs> {
+    return async (args): Promise<any> => { // TODO : TS
+      if (isAssistantEvent(args) && matchesConstraints(args)) return this.processEvent(args);
+      return args.next();
+    };
+  }
+
+  private async processEvent(args: AllAssistantMiddlewareArgs): Promise<void> {
+    const { payload } = args;
+    const assistantArgs = this.prepareAssistantArgs(args);
+    const assistantMiddleware = this.getAssistantMiddleware(payload);
+    return processAssistantMiddleware(assistantArgs, assistantMiddleware);
+  }
+
+  /**
+   * `getAssistantMiddleware()` returns the Assistant instance's middleware
+   */
+  private getAssistantMiddleware(payload: AllAssistantMiddlewareArgs['payload']): AssistantMiddleware {
+    switch (payload.type) {
+      case 'assistant_thread_started':
+        return this.threadStarted;
+      case 'assistant_thread_context_changed':
+        return this.threadContextChanged;
+      case 'message':
+        return this.userMessage;
+      default:
+        return [];
+    }
+  }
+
+  /**
+ * `prepareAssistantArgs()` takes the event arguments and:
+ *  1. Removes the next() passed in from App-level middleware processing, thus preventing
+ *  events from continuing down the global middleware chain to subsequent listeners
+ *  2. Adds assistant-specific utilities (i.e., helper methods)
+ * */
+  private prepareAssistantArgs(args: AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>): AllAssistantMiddlewareArgs {
+    const { next: _next, ...assistantArgs } = args;
+    const preparedArgs = { ...assistantArgs as Exclude<AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>, 'next'> };
+
+    preparedArgs.getThreadContext = () => this.threadContextStore.get(preparedArgs);
+    preparedArgs.saveThreadContext = () => this.threadContextStore.save(preparedArgs);
+    preparedArgs.say = createSay(preparedArgs);
+    preparedArgs.setStatus = createSetStatus(preparedArgs);
+    preparedArgs.setSuggestedPrompts = createSetSuggestedPrompts(preparedArgs);
+    preparedArgs.setTitle = createSetTitle(preparedArgs);
+
+    return preparedArgs;
```

Comment:
Sanity check: which utilities should be available for which methods?

`say`, for example, is currently available in `get` / `save` (doesn't work, but it could), but does it make sense to make them available there? Should any be made available to those methods?

Reply:
I don't think get/saveThreadContext should accept say argument. Once we release as-is and some developers use the say method inside the get/saveThreadContext method, removing say argument from the arg list in the future can be a breaking change

Reply:
Sounds good! I'll make sure `say` isn't made available in those methods.